forge/src/executor/inspector/cheatcodes/env.rs

```diff
@@ -0,0 +1,171 @@
+
+use super::Cheatcodes;
+use crate::abi::HEVMCalls;
+use bytes::Bytes;
+use ethers::{
+    abi::{AbiEncode, Token, Tokenize},
+    types::{Address, H256, U256},
+    utils::keccak256,
+};
+use revm::{Database, EVMData};
+
+#[derive(Clone, Debug, Default)]
+pub struct Prank {
+    /// Address of the contract that initiated the prank
+    pub prank_caller: Address,
+    /// Address of `tx.origin` when the prank was initiated
+    pub prank_origin: Address,
+    /// The address to assign to `msg.sender`
+    pub new_caller: Address,
+    /// The address to assign to `tx.origin`
+    pub new_origin: Option<Address>,
+    /// The depth at which the prank was called
+    pub depth: u64,
+    /// Whether or not the prank stops by itself after the next call
+    pub single_call: bool,
+}
+
+fn prank(
+    state: &mut Cheatcodes,
+    prank_caller: Address,
+    prank_origin: Address,
+    new_caller: Address,
+    new_origin: Option<Address>,
+    depth: u64,
+    single_call: bool,
+) -> Result<Bytes, Bytes> {
+    let prank = Prank {
+        prank_caller,
+        prank_origin,
+        new_caller,
+        new_origin,
+        // Note: When calling a cheatcode, the depth is not increased. Because of that, the depth
+        // we actually want to prank at is 1 level deeper.
+        depth: depth + 1,
+        single_call,
+    };
+
+    if state.prank.is_some() {
+        return Err("You have an active prank already.".to_string().encode().into())
+    }
+
+    state.prank = Some(prank);
+    Ok(Bytes::new())
+}
+
+#[derive(Clone, Debug, Default)]
+pub struct RecordAccess {
+    pub reads: BTreeMap<Address, Vec<U256>>,
+    pub writes: BTreeMap<Address, Vec<U256>>,
+}
+
+fn record(state: &mut Cheatcodes) {
+    state.accesses = Some(Default::default());
+}
+
+fn accesses(state: &mut Cheatcodes, address: Address) -> Bytes {
+    if let Some(storage_accesses) = &mut state.accesses {
+        ethers::abi::encode(&[
+            storage_accesses.reads.remove(&address).unwrap_or_default().into_tokens()[0].clone(),
+            storage_accesses.writes.remove(&address).unwrap_or_default().into_tokens()[0].clone(),
+        ])
+        .into()
+    } else {
+        ethers::abi::encode(&[Token::Array(vec![]), Token::Array(vec![])]).into()
+    }
+}
+
+pub fn apply<DB: Database>(
+    state: &mut Cheatcodes,
+    data: &mut EVMData<'_, DB>,
+    caller: Address,
+    call: &HEVMCalls,
+) -> Option<Result<Bytes, Bytes>> {
+    Some(match call {
+        HEVMCalls::Warp(inner) => {
+            data.env.block.timestamp = inner.0;
+            Ok(Bytes::new())
+        }
+        HEVMCalls::Roll(inner) => {
+            data.env.block.number = inner.0;
+            // TODO: Set blockhash somehow
+            Ok(Bytes::new())
+        }
+        HEVMCalls::Fee(inner) => {
+            data.env.block.basefee = inner.0;
+            Ok(Bytes::new())
+        }
+        HEVMCalls::Store(inner) => {
+            // TODO: Does this increase gas usage?
```

Comment:
I still haven't figured this out, will ping Dragan unless one of you know - whenever we want to change something about an account (nonce, storage, balance, ...) we need to call `load_account` first in the subroutine and it does a bunch of stuff. As far as I could tell there is no gas recording in there, it just returns some info about whether the account/slot/whatever was hot or cold, which I assume the rest of the REVM impl uses to account for gas.

Generally, though, there are some major issues with the current REVM port in terms of gas accounting, which I will address in a separate PR

Reply:
Sounds good - maybe @draganrakita can share some context here for anyone following the PR